FishTracker/tracker.py

```diff
@@ -0,0 +1,348 @@
+    :param all_times: (array) containing time stamps of frequency detection. (  len(all_times) == len(fishes[xy])  )
+    :param max_time_tolerance: (int) time in minutes from when a certain fish is no longer tracked.
+    :param freq_tolerance: (float) maximum frequency difference to assign a frequency to a certain fish.
+    :param save_original_fishes: (boolean) if True saves the sorted fishes after the first level of fish sorting.
+    :return fishes: (list) containing arrays of sorted fish frequencies. Each array represents one fish.
+    """
+    # ToDo: insert clean up function that deletes fishes with less than 5 detections after every hour...
+    detection_time_diff = np.median(np.diff(all_times))
+    dpm = 60. / detection_time_diff  # detections per minutes
+
+    fishes = [[ 0. ]]
+    last_fish_fundamentals = [ 0. ]
+    end_nans = [0]
+
+    # for every list of fundamentals ...
+    print len(all_fundamentals)
+    for t_list in range(len(all_fundamentals)):
+        # ... first add a nan to all fishes. replace later !!!
+        for fish in fishes:
+            fish.append(np.nan)
+
+        for idx in range(len(all_fundamentals[t_list])):
+            diff = abs(np.asarray(last_fish_fundamentals) - all_fundamentals[t_list][idx])
+            sorted_diff_idx = np.argsort(diff)
+            tollerated_diff_idx = sorted_diff_idx[diff[sorted_diff_idx] < freq_tolerance]
+
+            last_detect_of_tollerated = np.array(end_nans)[tollerated_diff_idx]
+
+            if len(tollerated_diff_idx) == 0:
+                fishes.append([np.nan for i in range(len(fishes[0]) - 1)])
+                fishes[-1].append(all_fundamentals[t_list][idx])
+                last_fish_fundamentals.append(all_fundamentals[t_list][idx])
+                end_nans.append(0)
+            else:
+                for i in tollerated_diff_idx[np.argsort(last_detect_of_tollerated)]:
+                # for i in tollerated_diff_idx:
+                    if np.isnan(fishes[i][-1]):
+                        fishes[i][-1] = all_fundamentals[t_list][idx]
+                        last_fish_fundamentals[i] = all_fundamentals[t_list][idx]
+                        end_nans[i] = 0
+                        break
+                    if i == tollerated_diff_idx[-1]:
+                        fishes.append([np.nan for i in range(len(fishes[0]) - 1)])
+                        fishes[-1].append(all_fundamentals[t_list][idx])
+                        last_fish_fundamentals.append(all_fundamentals[t_list][idx])
+                        end_nans.append(0)
+
+        for fish in range(len(end_nans)):
+            if end_nans[fish] >= max_time_tolerance * dpm:
+                last_fish_fundamentals[fish] = 0.
+
+        for fish in range(len(fishes)):
+            if np.isnan(fishes[fish][-1]):
+                end_nans[fish] += 1
+
+
+    # reshape everything to arrays
+    for fish in range(len(fishes)):
+        fishes[fish].pop(0)
```

Comment:
because it's not a real time point. Its an artifact from the beginning of the beginning of the beginning of the first_level_fish_sorting function where I compare the first fundamentals with as 0.